File: No_1413_Minimum Value to Get Positive Step by Step Sum/by_prefix_sum.py
'''

Description:

Given an array of integers nums, you start with an initial positive value startValue.

In each iteration, you calculate the step by step sum of startValue plus elements in nums (from left to right).

Return the minimum positive value of startValue such that the step by step sum is never less than 1.

 

Example 1:

Input: nums = [-3,2,-3,4,2]
Output: 5
Explanation: If you choose startValue = 4, in the third iteration your step by step sum is less than 1.
                step by step sum
                startValue = 4 | startValue = 5 | nums
                  (4 -3 ) = 1  | (5 -3 ) = 2    |  -3
                  (1 +2 ) = 3  | (2 +2 ) = 4    |   2
                  (3 -3 ) = 0  | (4 -3 ) = 1    |  -3
                  (0 +4 ) = 4  | (1 +4 ) = 5    |   4
                  (4 +2 ) = 6  | (5 +2 ) = 7    |   2


                  
Example 2:

Input: nums = [1,2]
Output: 1
Explanation: Minimum start value should be positive. 



Example 3:

Input: nums = [1,-2,-3]
Output: 5
 

Constraints:

1 <= nums.length <= 100
-100 <= nums[i] <= 100

'''



from typing import List

class Solution:
    def minStartValue(self, nums: List[int]) -> int:
        
        prefix_sum = 0
        negative_most_offset = float('inf')
        
        # linear scan, and update nagative most offset by prefix sum
        for i in range(0, len(nums)):
            
            prefix_sum += nums[i]
            negative_most_offset = min( prefix_sum, negative_most_offset)
        
        # compute the minimum value to get positive
        threshold = min( negative_most_offset,  0 )
        
        # abs to flip the sign, +1 to make it larger than zero
        return abs(threshold)+1



## time complexity: O( n )
#
# The overhead in time isthe cost of linear scna, which is of O( n ).

## Space Complexity: O( 1 )
#
# The overhead in space is the storage for loop index and temporary variable for computation, which are of O( n ).


from collections import namedtuple
TestEntry = namedtuple('TestEntry', 'sequence')

def test_bench():

    test_data = [
                    TestEntry( sequence = [-3,2,-3,4,2] ),
                    TestEntry( sequence = [1,2] ),
                    TestEntry( sequence = [1,-2,-3] ),
                    TestEntry( sequence = [-5] ),
                    TestEntry( sequence = [9] ),
                ]        

    # expected output:
    '''
    5
    1
    5
    6
    1
    '''

    for t in test_data:

        print( Solution().minStartValue( nums = t.sequence) )

    return



if __name__ == '__main__':     
    
    test_bench()

File: No_1074_Number of Submatrices That Sum to Target/by_prefix_sum.go
/*

Description:

Given a matrix and a target, return the number of non-empty submatrices that sum to target.

A submatrix x1, y1, x2, y2 is the set of all cells matrix[x][y] with x1 <= x <= x2 and y1 <= y <= y2.

Two submatrices (x1, y1, x2, y2) and (x1', y1', x2', y2') are different if they have some coordinate that is different: for example, if x1 != x1'.

 

Example 1:


Input: matrix = [[0,1,0],[1,1,1],[0,1,0]], target = 0
Output: 4
Explanation: The four 1x1 submatrices that only contain 0.



Example 2:

Input: matrix = [[1,-1],[-1,1]], target = 0
Output: 5
Explanation: The two 1x2 submatrices, plus the two 2x1 submatrices, plus the 2x2 submatrix.



Example 3:

Input: matrix = [[904]], target = 0
Output: 0
 

Constraints:

1 <= matrix.length <= 100
1 <= matrix[0].length <= 100
-1000 <= matrix[i] <= 1000
-10^8 <= target <= 10^8

*/

package No_1074

func numSubmatrixSumTarget(matrix [][]int, target int) int {
    
    // height and width of matrix
    h, w := len(matrix), len(matrix[0])
    
    // update prefix sum on each row
    for y := 0 ; y < h ; y++{
        for x:= 1 ; x < w; x++{
            matrix[y][x] = matrix[y][x] + matrix[y][x-1]
        }
    }
    
    // number of submatrices that sum to target
    counter := 0
    
    // sliding windows on x-axis, in range[left, right]
    for left := 0 ; left < w ; left ++{
        for right := left ; right < w ; right++{
            
            // accumulation for area so far
            accumulation := map[int]int{ 0 : 1}
            
            // area of current submatrices, bounded by [left, right] with height y
            area := 0
            
            // scan each possible height on y-axis
            for y := 0 ; y < h ; y++{
                
                if left > 0 {
                    area += matrix[y][right] - matrix[y][left-1]
                }else{
                    area += matrix[y][right]
                }
                
                // if ( area - target ) exists, then target must exist in submatrices
                counter += accumulation[ (area - target) ]
                
                // update dictionary with current accumulation area
                accumulation[area] += 1
                
            }
            
        }//end of right loop
        
    }//end of left loop
    
    return counter
    
}


// w : width of input matrix
// h : height of input matrix

//// Time Complexity: O( w^2 * h )
//
// The overhead in time is the cost of iteration, which is of O( w^2 * h )

//// Space Complexity: O( w * h )
//
// The overhead in space is the storage for dictionary, which is of O( w * h )

